types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,189 @@
+package mempool
+
+import (
+	crand "crypto/rand" // #nosec // crypto/rand is used for seed generation
+	"encoding/binary"
+	"fmt"
+	"math/rand" // #nosec // math/rand is used for random selection and seeded from crypto/rand
+
+	huandu "github.com/huandu/skiplist"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+var (
+	_ Mempool  = (*senderNonceMempool)(nil)
+	_ Iterator = (*senderNonceMepoolIterator)(nil)
+)
+
+type senderNonceMempool struct {
+	senders map[string]*huandu.SkipList
+	rnd     *rand.Rand
+}
+
+// NewSenderNonceMempool creates a new mempool that prioritizes transactions by nonce, the lowest first.
+func NewSenderNonceMempool() Mempool {
+	senderMap := make(map[string]*huandu.SkipList)
+	snp := &senderNonceMempool{
+		senders: senderMap,
+	}
+
+	var seed int64
+	err := binary.Read(crand.Reader, binary.BigEndian, &seed)
+	if err != nil {
+		panic(err)
+	}
+	snp.setSeed(seed)
+
+	return snp
+}
+
+// NewSenderNonceMempoolWithSeed creates a new mempool that prioritizes transactions by nonce, the lowest first and sets the random seed.
+func NewSenderNonceMempoolWithSeed(seed int64) Mempool {
+	senderMap := make(map[string]*huandu.SkipList)
+	snp := &senderNonceMempool{
+		senders: senderMap,
+	}
+	snp.setSeed(seed)
+	return snp
+}
+
+func (snm *senderNonceMempool) setSeed(seed int64) {
+	s1 := rand.NewSource(seed)
+	snm.rnd = rand.New(s1) //#nosec // math/rand is seeded from crypto/rand by default
+}
+
+// Insert adds a tx to the mempool. It returns an error if the tx does not have at least one signer.
+// priority is ignored.
+func (snm *senderNonceMempool) Insert(_ sdk.Context, tx sdk.Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	senderTxs, found := snm.senders[sender]
+	if !found {
+		senderTxs = huandu.New(huandu.Uint64)
+		snm.senders[sender] = senderTxs
+	}
+	senderTxs.Set(nonce, tx)
+
+	return nil
+}
+
+// Select returns an iterator ordering transactions the mempool with the lowest nonce of a random selected sender first.
+func (snm *senderNonceMempool) Select(_ sdk.Context, _ [][]byte) Iterator {
+	var senders []string
+	senderCursors := make(map[string]*huandu.Element)
+	// #nosec
+	for key := range snm.senders {
+		senders = append(senders, key)
+		senderCursors[key] = snm.senders[key].Front()
+	}
+
+	iter := &senderNonceMepoolIterator{
+		senders:       senders,
+		rnd:           snm.rnd,
+		senderCursors: senderCursors,
+	}
```

Comment:
if sender A submits a tx before sender B, then sender b **could** be executed before sender A's tx right?

Reply:
right

Reply:
Correct

Reply:
awesome thanks. could we add a short doc.go explaining this. can be a few sentences. helps needing to send people to read code.